core/src/impl/Kokkos_Default_GraphNode_Impl.hpp

```diff
@@ -92,6 +92,18 @@ struct GraphNodeBackendSpecificDetails {
     m_is_aggregate = true;
   }
 
+  // A node is awaitable if it can execute a kernel.
+  // A root node or an aggragate node cannot be waited for, because it does
+  // not launch anything.
+  bool awaitable() const { return (!m_is_root) && (!m_is_aggregate); }
+
+  // Retrieve the execution space instance that has been passed to
+  // the kernel at construction phase.
+  const ExecutionSpace& get_execution_space() const {
+    KOKKOS_EXPECTS(m_kernel_ptr != nullptr && awaitable())
```

Comment:
Let me try to summarize this.

Looking at `Kokkos::Impl::GraphNodeImpl`, it inherits from `Kokkos::Impl::GraphNodeBackendSpecificDetails` and `Kokkos::Impl::ExecutionSpaceInstanceStorage`. So it could be tempting to do something like:
```c++
for (const auto& sink : m_sinks) {
  if (sink->awaitable()) {
    const auto & tmp = static_cast<GraphNodeImpl*>(sink.get())->execution_space_instance();
    tmp.fence("fencing before ending graph submit");
  }
}
```
and get rid of the things I've added.

The thing is, it won't do what I want.

It seems that `Kokkos::Impl::GraphNodeImpl::execution_space_instance` contains the execution space instance given to the `create_graph`.

And since our beloved kernel nodes will launch on the execution space instance of their execution policy, which i explicitly made different from the execution space instance passed to the `create_graph`, you see the problem coming.

So I had to sort of "catch" the execution space instance of the kernel execution policy `GraphNodeKernelImpl`, which is only used for kernel nods, not for root nor aggregates.

Reply:
Now, answering your question, I think `m_kernel_impl` will be `nullptr` if the node is not `awaitable`, *i.e.* it should be `nullptr` for root and aggregates. So probably `KOKKOS_EXPECTS(m_kernel_ptr != nullptr)` would be sufficient.

Reply:
That's what I was thinking

Reply:
OK, I will change to `KOKKOS_EXPECTS(m_kernel_ptr != nullptr)`.

Reply:
@dalg24 OK to close this thread ?